src/Raven.Server/Documents/ResourceCache.cs

```diff
@@ -86,44 +100,63 @@ private bool UnlikelyTryGet(StringSegment resourceName, out Task<TResource> reso
 
         public bool TryRemove(StringSegment resourceName, Task<TResource> resourceTask)
         {
-            if (_caseInsensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(resourceName, resourceTask)) == false)
-                return false;
-
-            _resourceDetails.TryRemove(resourceTask, out _);
-
             lock (this)
             {
-                RemoveCaseSensitive(resourceName, resourceTask);
-            }
+                if (_caseInsensitive.TryRemove(new KeyValuePair<StringSegment, Task<TResource>>(resourceName, resourceTask)) == false)
+                    return false;
 
-            return true;
-        }
+                _resourceDetails.Remove(resourceTask, out _);
 
-        public bool TryGetAndRemove(StringSegment resourceName, out Task<TResource> resourceTask)
-        {
-            if (TryGetValue(resourceName, out resourceTask) == false)
-                return false;
+                RemoveCaseSensitive(resourceName, resourceTask);
 
-            if (TryRemove(resourceName, resourceTask) == false)
-                return false;
+                // We need to refresh all of them. Since construction takes time, we will update all the references
+                // at the end of the execution in order to diminish the time that an unstable state can be observed.
+                // While this is not a big issue because the code already supports dealing with that, the smaller the
+                // time such inconsistencies can be observed, the better.
+                var readonlyCaseSensitive = _caseSensitive.ToFrozenDictionary();
+                var readonlyCaseInsensitive = _caseInsensitive.ToFrozenDictionary();
+                var readonlyResourceDetails = _resourceDetails.ToFrozenDictionary();
+
```

Comment:
I guess we code wrap this in a Tuple, and replace that atomically, That would require additional indirection always, probably not a good idea, I'm assuming

Reply:
@redknightlois  ?

Reply:
Not really  a big deal, the code itself handles this explicitly. There is no upside to do that and the code becomes much less legible.